From 1f5b9edd73039f461a80275a58b97bca3739c948 Mon Sep 17 00:00:00 2001
From: Josh Wu <josh.wu@atmel.com>
Date: Wed, 12 Feb 2014 16:38:07 +0800
Subject: [PATCH] ARM: sama5d3: enable mother board & display module board
 detect

Detect mother board revision and create a environment variable: mb_rev
  md_rev can be 'a', 'b', 'c', 'd' and etc.

Detect display module type and create a environment variable: dm_type
  1. if display module is generate touch screen then dm_typ is empty.
  2. for PDA display module the dm_type can be 'pda4' or 'pda7'.

Signed-off-by: Josh Wu <josh.wu@atmel.com>
---
 board/atmel/sama5d3xek/sama5d3xek.c | 71 +++++++++++++++++++++++++++++++++++++
 include/configs/sama5d3xek.h        |  3 ++
 2 files changed, 74 insertions(+)

diff --git a/board/atmel/sama5d3xek/sama5d3xek.c b/board/atmel/sama5d3xek/sama5d3xek.c
index 67b7bb9..e0e9d64 100644
--- a/board/atmel/sama5d3xek/sama5d3xek.c
+++ b/board/atmel/sama5d3xek/sama5d3xek.c
@@ -35,6 +35,45 @@ DECLARE_GLOBAL_DATA_PTR;
  * Miscelaneous platform dependent initialisations
  */
 
+/* One-wire information detect */
+
+/*
+ * SN layout
+ *   31  30         25         20         15         10         5          0
+ *   -----------------------------------------------------------------------
+ *   |   | Vendor ID| Board ID | Vendor ID| Board ID | Vendor ID| Board ID |
+ *   -----------------------------------------------------------------------
+ *       |         EK          |         DM          |         CPU         |
+ */
+#define GPBR_ONEWIRE_INFO_SN	2	/* SN info is in GPBR2 */
+#define		BOARD_ID_SAM9x5_DM	1
+#define		BOARD_ID_PDA_DM		7
+#define		BOARD_ID_SAMA5D3X_DM	9
+#define		BOARD_ID_PDA7_DM	15
+char get_dm_board_id(void)
+{
+	return (readl(ATMEL_BASE_GPBR + 4 * GPBR_ONEWIRE_INFO_SN) >> 10) & 0x1f;
+}
+
+/*
+ * REV layout
+ *   31              24     21     18     15         10         5          0
+ *   -----------------------------------------------------------------------
+ *   |               |  EK  |  DM  |  CPU |    EK    |    DM    |   CPU    |
+ *   -----------------------------------------------------------------------
+ *                   |     Revision id    |        Revision Code           |
+ */
+#define GPBR_ONEWIRE_INFO_REV	3	/* REV info is in GPBR3 */
+char get_mb_rev_code(void)
+{
+	char mb_rev_code = 'a';
+
+	mb_rev_code += 0x1f &
+		(readl(ATMEL_BASE_GPBR + 4 * GPBR_ONEWIRE_INFO_REV) >> 10);
+
+	return mb_rev_code;
+}
+
 #ifdef CONFIG_NAND_ATMEL
 void sama5d3xek_nand_hw_init(void)
 {
@@ -302,6 +341,38 @@ void spi_cs_deactivate(struct spi_slave *slave)
 }
 #endif /* CONFIG_ATMEL_SPI */
 
+#ifdef CONFIG_BOARD_LATE_INIT
+int board_late_init(void)
+{
+#ifdef CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG
+	const char *SAMA5D3_MB_REV_ENV_NAME = "mb_rev";
+	const char *SAMA5D3_DM_TYPE_ENV_NAME = "dm_type";
+	char rev_code[2], dm_id;
+
+	rev_code[0] = get_mb_rev_code();
+	rev_code[1] = '\0';
+	setenv(SAMA5D3_MB_REV_ENV_NAME, rev_code);
+
+	dm_id = get_dm_board_id();
+	switch (dm_id) {
+	case BOARD_ID_PDA_DM:
+		setenv(SAMA5D3_DM_TYPE_ENV_NAME, "pda4");
+		break;
+	case BOARD_ID_PDA7_DM:
+		setenv(SAMA5D3_DM_TYPE_ENV_NAME, "pda7");
+		break;
+	case BOARD_ID_SAM9x5_DM:
+	case BOARD_ID_SAMA5D3X_DM:
+	default:
+		setenv(SAMA5D3_DM_TYPE_ENV_NAME, "");
+		break;
+	}
+#endif
+
+	return 0;
+}
+#endif
+
 /* SPL */
 #ifdef CONFIG_SPL_BUILD
 void spl_board_init(void)
diff --git a/include/configs/sama5d3xek.h b/include/configs/sama5d3xek.h
index 07e1c80..2127213 100644
--- a/include/configs/sama5d3xek.h
+++ b/include/configs/sama5d3xek.h
@@ -35,6 +35,9 @@
 
 #define CONFIG_SYS_GENERIC_BOARD
 
+#define CONFIG_BOARD_LATE_INIT
+#define CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG
+
 /* general purpose I/O */
 #define CONFIG_AT91_GPIO
 
-- 
1.9.1

